### PR TITLE
Improve RenderShroudCircle

### LINKED ISSUE #15393: RenderShroudCircle has no IsTraitDisabled check
RenderShroudCircle is always active regardless of CreatesShroud's conditional status. It should behave like RenderDetectionCircle and only be active when the trait is enabled.

### PR DESCRIPTION
**Pull request by chrisforbes**
Also align this implementation with other circles.

- Unhardcoded colors
- support for multiple CreatesShroud traits
- cache range at actor creation

### CODE REVIEW

**Review by Mailaender — DISMISSED** [review]

File: OpenRA.Mods.Cnc/Traits/Render/RenderShroudCircle.cs [hunk 1]
```diff
@@ -19,37 +19,60 @@
...
+		{
+			this.info = info;
+		}
+
 		public IEnumerable<IRenderable> RangeCircleRenderables(Actor self, WorldRenderer wr)
 		{
 			if (!self.Owner.IsAlliedWith(self.World.RenderPlayer))
 				yield break;
 
+			var range = self.TraitsImplementing<CreatesShroud>()
```
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Traits/Render/RenderShroudCircle.cs]: This can be cached in the ctor.

**Review by abcdefg30 — COMMENTED** [review]
Looks good to me otherwise.

File: OpenRA.Mods.Cnc/Traits/Render/RenderShroudCircle.cs (hunk 1 above)
Comment (GraionDilach) [reply OpenRA.Mods.Cnc/Traits/Render/RenderShroudCircle.cs]: Not in the ctor, since there is no Requires<>, but in INotifyCreated.

**Review by pchote — COMMENTED** [review]
I have force pushed over this to rebase and cache the enumerable.

The code LGTM, but i'm not convinced about the functionality: we don't disable the range circles on low-power base defenses, so why disable them on the gap generator?

**Comment by reaperrr** [comment]
> The code LGTM, but i'm not convinced about the functionality: we don't disable the range circles on low-power base defenses, so why disable them on the gap generator?

It would be a shame for the other improvements to stay unmerged over this, so how about adding a `ShowDisabledShroudRanges` boolean with the corresponding alternate caching, and let it default to `true` for now?

**Comment by pchote** [comment]
How about we drop the disabled check so that we can merge the other improvements without changing the behaviour?  Discussion can then continue in #15393.

**Comment by pchote** [comment]
Ok, done. :+1: on my part, but lets get another pair of eyes on my changes.

**Review by pchote — DISMISSED** [review]

File: OpenRA.Mods.Cnc/Traits/Render/RenderShroudCircle.cs [hunk 2]
```diff
@@ -19,37 +19,61 @@
...
+		public readonly Color Color = Color.FromArgb(128, Color.Cyan);
+
+		[Desc("Contrast color of the circle.")]
+		public readonly Color ContrastColor = Color.FromArgb(96, Color.Black);
+
 		public IEnumerable<IRenderable> Render(WorldRenderer wr, World w, ActorInfo ai, WPos centerPosition)
 		{
 			var localRange = new RangeCircleRenderable(
 				centerPosition,
 				ai.TraitInfo<CreatesShroudInfo>().Range,
```
Comment (reaperrr) [inline OpenRA.Mods.Cnc/Traits/Render/RenderShroudCircle.cs]: This actually needs to be updated as well, else it crashes when more than 1 CreatesShroud trait is present.

**Comment by pchote** [comment]
Pushed a rebase and update.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/364fb548df91e0d2100e7f3b04c205b907ded907...f87911c699926ffd5045b3505cf98e28ea19a04f)

**Pull request merged**